qualtran/drawing/qpic_diagram.py

```diff
@@ -0,0 +1,332 @@
+        controls, targets = [], []
+
+        def add_soq(soq: Soquet):
+            wire, gate = self._add_soq(soq)
+            if gate is None:
+                controls.append(wire)
+            else:
+                targets.append(wire + gate)
+
+        for cxn in pred:
+            self.add_connection(cxn)
+            add_soq(cxn.right)
+
+        for cxn in succ:
+            self.add_connection(cxn)
+            if cxn.left.reg.side == Side.RIGHT:
+                add_soq(cxn.left)
+        self.gates += [' '.join(targets + controls)]
+
+
+def get_qpic_data(bloq: 'Bloq', file_path: Union[None, pathlib.Path, str] = None) -> List[str]:
+    """Get the input data that can be used to draw a latex diagram for `bloq` using `qpic`.
+
+    Args:
+        bloq: Bloqs to draw the latex diagram for
+        file_path: If specified, the output is stored at the file.
+
+    Returns:
+        A list of strings representing the input to `qpic` utility to draw a latex diagram
+        for bloq. If specified, the output is written to `file_path`.
+    """
+    cbloq = bloq.as_composite_bloq()
+    qpic_circuit = QpicCircuit(cbloq.signature)
+    for binst, pred, succ in cbloq.iter_bloqnections():
+        qpic_circuit.add_bloq(binst.bloq, pred, succ)
+    data = qpic_circuit.wires + qpic_circuit.gates
+    if file_path:
+        with open(file_path, 'w') as f:
+            f.write('\n'.join(data))
+    else:
+        return data
+
+
+def _to_snake_case(name):
+    """Convert camel case to snake case. Taken from https://stackoverflow.com/a/1176023"""
+    name = re.sub('(.)([A-Z][a-z]+)', r'\1_\2', name)
+    name = re.sub('__([A-Z])', r'_\1', name)
+    name = re.sub('([a-z0-9])([A-Z])', r'\1_\2', name)
+    return name.lower()
+
+
+def qpic_input_to_diagram(
+    qpic_file: Union[pathlib.Path, str],
+    output_file: Union[None, pathlib.Path, str] = None,
+    output_type: str = 'pdf',
+) -> str:
+    r"""Invoke `qpic` script to generate output diagram of type qpic/tex/pdf/png.
+
+    Outputs one of the following files, based on `output_type`:
+     - base_file_path + '.qpic' - Copies the input qpic_file to output_file.
+     - base_file_path + '.tex' - Obtained via `qpic -f base_file_path.qpic`
+     - base_file_path + '.pdf' - Uses `pdflatex` tool to convert tex to pdf. See
+        https://tug.org/applications/pdftex/ for more details on how to install `pdflatex`.
+     - base_file_path + '.png' - Uses `convert` tool to convert pdf to png. See
+        https://imagemagick.org/ for more details on how to install `convert`.
+
+    Args:
+        qpic_file: Path to file containing input that should be passed to the `qpic` script.
+        output_file: Optional path to the output where generated diagram should be saved. Defaults to
+            qpic_file.with_suffix(f".{output_type}")
+        output_type: Format of the diagram generated using qpic. Supported output types are one of
+            ['qpic', 'tex', 'pdf', 'png']
```

Comment:
what happens if output_file extension and output_type conflict?

Reply:
I use temporary files with the right extension as intermediate files to store results and move them to the output file when the right file is computed, so it doesn't matter what the extension of the output file is